File: app/javascript/channels/multiplayer_channel.js
import consumer from "channels/consumer";

let multiplayerChannel = null;

// Function to initialize the MultiplayerChannel subscription
export function initializeMultiplayerChannel(gridId, onPlayerUpdate) {
  if (!gridId) {
    console.error("Grid ID is required to initialize MultiplayerChannel.");
    return;
  }

  // Unsubscribe from any existing channel
  if (multiplayerChannel) {
    multiplayerChannel.unsubscribe();
    multiplayerChannel = null;
  }

  // Subscribe to the MultiplayerChannel for the given grid
  multiplayerChannel = consumer.subscriptions.create(
      { channel: "MultiplayerChannel", grid_id: gridId },
      {
        received(data) {
          console.log("Received multiplayer update:", data);

          // Handle different types of updates
          if (data.type === "player_joined") {
            onPlayerUpdate("joined", data.username, data.cell_id);
          } else if (data.type === "update_position") {
            onPlayerUpdate("moved", data.username, data.cell_id);
          } else if (data.type === "player_left") {
            onPlayerUpdate("left", data.username);
          }
        },
        connected() {
          console.log(`Connected to MultiplayerChannel for grid ${gridId}`);
        },
        disconnected() {
          console.log("Disconnected from MultiplayerChannel.");
        }
      }
  );
}

// Function to update the current player's position
export function updatePlayerPosition(cellId) {
  if (!multiplayerChannel) {
    console.error("MultiplayerChannel is not initialized.");
    return;
  }

  multiplayerChannel.perform("update_position", { cell_id: cellId });
}

// Function to unsubscribe from the MultiplayerChannel
export function unsubscribeMultiplayerChannel() {
  if (multiplayerChannel) {
    multiplayerChannel.unsubscribe();
    multiplayerChannel = null;
    console.log("Unsubscribed from MultiplayerChannel.");
  }
}

File: app/channels/multiplayer_channel.rb
class MultiplayerChannel < ApplicationCable::Channel
  def subscribed
    character = current_user.character
    character.update!(online_status: true)

    stream_from "multiplayer_grid_#{character.grid_id}"

    # Transmit the status of all players in the current grid
    online_players = Character.where(grid_id: character.grid_id, online_status: true).map do |player|
      { character_name: player.character_name, cell_id: player.cell_id }
    end
    transmit({ type: "grid_state", players: online_players })

    # Notify other players that a new player has joined
    ActionCable.server.broadcast("multiplayer_grid_#{character.grid_id}", {
      type: "player_joined",
      character: character.broadcast_data
    })
  end

  def unsubscribed
    character = current_user.character
    character.update!(online_status: false)

    ActionCable.server.broadcast("multiplayer_grid_#{character.grid_id}", {
      type: "player_left",
      character: {
        character_name: character.character_name
      }
    })
  end

  def update_position(data)
    character = current_user.character
    character.update!(cell_id: data["cell_id"])

    # Log before broadcast
    logger.info "Broadcasting position update: #{character.character_name} moved to #{data["cell_id"]}"

    # Broadcast the player's new location
    ActionCable.server.broadcast("multiplayer_grid_#{character.grid_id}", {
      type: "update_position",
      character: {
        character_name: character.character_name,
        cell_id: character.cell_id
      }
    })
  end
end
